### ISSUE TITLE
[Jepsen] [YSQL] Inconsistency in bank test

### ISSUE THREAD

**Issue by frozenspider** [YugaByte/yugabyte-db#1553]
(Does not affect YCQL)
We're observing inconsistency in YSQL banking test (transactionally moves funds around from one account to another) - even without any nemeses funds total diverge from initial $100

Full log: [logs-jepsen_2019-06-17_15-42-51.tar.gz](https://github.com/YugaByte/yugabyte-db/files/3296488/logs-jepsen_2019-06-17_15-42-51.tar.gz)

Reproducible with master 6d65f74, related to #938

**Comment by bmatican** [comment]
For reference, we have some c++ level bank test as well, but on snapshot level: `src/yb/client/snapshot-txn-test.cc`:
```
277
278 TEST_F(SnapshotTxnTest, BankAccounts) {
279   TestBankAccounts({}, 30s, RegularBuildVsSanitizers(10, 1) /* minimal_updates_per_second */);
280 }
281
282 TEST_F(SnapshotTxnTest, BankAccountsWithTimeStrobe) {
283   TestBankAccounts(
284       BankAccountsOptions{BankAccountsOption::kTimeStrobe}, 300s,
285       RegularBuildVsSanitizers(10, 1) /* minimal_updates_per_second */);
286 }
287
288 TEST_F(SnapshotTxnTest, BankAccountsWithTimeJump) {
289   TestBankAccounts(
290       BankAccountsOptions{BankAccountsOption::kTimeJump, BankAccountsOption::kStepDown}, 30s,
291       RegularBuildVsSanitizers(3, 1) /* minimal_updates_per_second */);
292 }
```

**Comment by frozenspider** [comment]
Interesting excerpt from Jepsen history ([logs-jepsen_2019-07-03_15-30-38__bank-w-logs-1.tar.gz](https://github.com/YugaByte/yugabyte-db/files/3354960/logs-jepsen_2019-07-03_15-30-38__bank-w-logs-1.tar.gz))
```
...
{:type :ok, :f :read, :op-index 36, :process 8, :time 29378788717, :value {0 94, 1 9, 2 1, 3 4, 4 3, 5 -3, 6 -4, 7 -4}, :op-timing ["2019-07-03_12:19:041.656" "2019-07-03_12:19:041.686"], :index 62}
{:type :invoke, :f :read, :op-index 37, :process 8, :time 29380048765, :index 63}
{:type :ok, :f :transfer, :value {:from 1, :to 4, :amount 1}, :op-index 29, :process 9, :time 29431936282, :op-timing ["2019-07-03_12:19:040.535" "2019-07-03_12:19:041.740"], :index 64}
{:type :invoke, :f :transfer, :value {:from 2, :to 6, :amount 4}, :op-index 38, :process 9, :time 29432885187, :index 65}
{:type :ok, :f :read, :op-index 37, :process 8, :time 29464765889, :value {0 94, 1 9, 2 1, 3 4, 4 4, 5 -3, 6 -4, 7 -4}, :op-timing ["2019-07-03_12:19:041.688" "2019-07-03_12:19:041.772"], :index 66}
...
```
First read with `:op-index 36` is OK (sums up to 100), then transfer `:op-index 29` which (moves $1 from acct 1 to acct 4) returns OK.
The next read `:op-index 37` shows that acct 4 received $1 but it wasn't subtracted from acct 1

**Comment by frozenspider** [comment]
Update: this specific case was probably due to the fact that `read` op wasn't explicitly wrapped in transaction (and thus used snapshot isolation)

**Comment by frozenspider** [comment]
Confirmed: This behaviour (not observing all of the updated introduced by committed transactions) is only reproducible for `SNAPSHOT`-level reads.
Once I wrapped read in explicit transaction, this issue stopped reproducing.

**Comment by frozenspider** [comment]
Created a follow-up ticket: https://github.com/YugaByte/yugabyte-db/issues/1688

**Comment by frozenspider** [comment]
As it turns out, this test still has issues when ran with high `--time-limit`:
[logs-jepsen_2019-07-05_16-50-21__bank_c10_tl600.tar.gz](https://github.com/YugaByte/yugabyte-db/files/3362945/logs-jepsen_2019-07-05_16-50-21__bank_c10_tl600.tar.gz)

**Comment by frozenspider** [comment]
No longer reproducible after @spolitov's fix to #1577 in commit 76fcf86
